src/main/scala/firrtl/transforms/DeadCodeElimination.scala

```diff
@@ -0,0 +1,251 @@
+        getDeps(expr).foreach(ref => depGraph.addEdge(node, ref))
+      // Simulation constructs are treated as top-level outputs
+      case Stop(_,_, clk, en) =>
+        Seq(clk, en).flatMap(getDeps(_)).foreach(ref => depGraph.addEdge(circuitSink, ref))
+      case Print(_, _, args, clk, en) =>
+        (args :+ clk :+ en).flatMap(getDeps(_)).foreach(ref => depGraph.addEdge(circuitSink, ref))
+      case Block(stmts) => stmts.foreach(onStmt(_))
+      case ignore @ (_: IsInvalid | _: WDefInstance | EmptyStmt) => // do nothing
+      case other => throw new Exception(s"Unexpected Statement $other")
+    }
+
+    // Add all ports as vertices
+    mod.ports.foreach { case Port(_, name, _, _: GroundType) =>
+      depGraph.addVertex(LogicNode(mod.name, name))
+    }
+    onStmt(mod.body)
+  }
+
+  // TODO Make immutable?
+  private def createDependencyGraph(instMaps: collection.Map[String, collection.Map[String, String]],
+                                    c: Circuit): MutableDiGraph[LogicNode] = {
+    val depGraph = new MutableDiGraph[LogicNode]
+    c.modules.foreach {
+      case mod: Module => setupDepGraph(depGraph, instMaps(mod.name))(mod)
+      case ext: ExtModule =>
+        // Connect all inputs to all outputs
+        val node = LogicNode(ext.name, ext.name)
+        ext.ports.foreach {
+          case Port(_, pname, _, AnalogType(_)) =>
+            depGraph.addEdge(LogicNode(ext.name, pname), node)
+            depGraph.addEdge(node, LogicNode(ext.name, pname))
+          case Port(_, pname, Output, _) => depGraph.addEdge(LogicNode(ext.name, pname), node)
+          case Port(_, pname, Input, _) => depGraph.addEdge(node, LogicNode(ext.name, pname))
+        }
+    }
+    // Connect circuitSink to top-level outputs (and Analog inputs)
+    val topModule = c.modules.find(_.name == c.main).get
+    val topOutputs = topModule.ports.foreach {
+      case Port(_, name, Output, _) => depGraph.addEdge(circuitSink, LogicNode(c.main, name))
+      case Port(_, name, _, AnalogType(_)) => depGraph.addEdge(circuitSink, LogicNode(c.main, name))
+      case _ =>
+    }
+
+    depGraph
+  }
+
+  private def deleteDeadCode(instMap: collection.Map[String, String],
+                             deadNodes: Set[LogicNode],
+                             moduleMap: collection.Map[String, DefModule])
+                            (mod: DefModule): Option[DefModule] = {
+    // Gets all dependencies and constructs LogicNodes from them
+    // TODO this is a duplicate from setupDepGraph, remove by improving how we lookup
+    def getDeps(expr: Expression): Seq[LogicNode] =
+      extractRefs(expr).map { e =>
+        if (kind(e) == InstanceKind) {
+          val (inst, tail) = Utils.splitRef(e)
+          LogicNode(instMap(inst.name), tail)
+        } else {
+          LogicNode(mod.name, e)
+        }
+      }
+
+    var emptyBody = true // TODO is there a better way to do this?
+
+    // TODO Delete unused writers from DefMemory???
```

Comment:
Should probably be moved to DefMemory in setupDepGraph - also, I don't think you can dce write ports without optimizing the entire memory away, but you can dce read ports.